storage/src/test/java/org/apache/kafka/storage/internals/log/LogSegmentTest.java

```diff
@@ -0,0 +1,763 @@
+        // Given two messages with a gap between them (e.g. mid offset compacted away)
+        MemoryRecords ms1 = recordsForTruncateEven(offset, "first message");
+        seg.append(offset, RecordBatch.NO_TIMESTAMP, -1L, ms1);
+        MemoryRecords ms2 = recordsForTruncateEven(offset + 3, "message after gap");
+        seg.append(offset + 3, RecordBatch.NO_TIMESTAMP, -1L, ms2);
+
+        // When we truncate to an offset without a corresponding log entry
+        seg.truncateTo(offset + 1);
+
+        // Then we should still truncate the record that was present (i.e. offset + 3 is gone)
+        FetchDataInfo log = seg.read(offset, 10000);
+        assertEquals(offset, log.records.batches().iterator().next().baseOffset());
+        assertEquals(1, iteratorToList(log.records.batches().iterator()).size());
+    }
+
+    private MemoryRecords records(long offset, int size) {
+        return MemoryRecords.withRecords(RecordBatch.MAGIC_VALUE_V2, offset, Compression.NONE, TimestampType.CREATE_TIME,
+            new SimpleRecord(new byte[size]));
+    }
+
+    @Test
+    public void testAppendFromFile() throws IOException {
+        // create a log file in a separate directory to avoid conflicting with created segments
+        File tempDir = TestUtils.tempDirectory();
+        FileRecords fileRecords = FileRecords.open(LogFileUtils.logFile(tempDir, 0));
+
+        // Simulate a scenario with log offset range exceeding Integer.MAX_VALUE
+        fileRecords.append(records(0, 1024));
+        fileRecords.append(records(500, 1024 * 1024 + 1));
+        long sizeBeforeOverflow = fileRecords.sizeInBytes();
+        fileRecords.append(records(Integer.MAX_VALUE + 5L, 1024));
+        long sizeAfterOverflow = fileRecords.sizeInBytes();
+
+        LogSegment segment = createSegment(0);
+        long bytesAppended = segment.appendFromFile(fileRecords, 0);
+        assertEquals(sizeBeforeOverflow, bytesAppended);
+        assertEquals(sizeBeforeOverflow, segment.size());
+
+        LogSegment overflowSegment = createSegment(Integer.MAX_VALUE);
+        long overflowBytesAppended = overflowSegment.appendFromFile(fileRecords, (int) sizeBeforeOverflow);
+        assertEquals(sizeAfterOverflow - sizeBeforeOverflow, overflowBytesAppended);
+        assertEquals(overflowBytesAppended, overflowSegment.size());
+
+        Utils.delete(tempDir);
+    }
+
+    @Test
+    public void testGetFirstBatchTimestamp() throws IOException {
+        LogSegment segment = createSegment(1);
+        assertEquals(Long.MAX_VALUE, segment.getFirstBatchTimestamp());
+
+        segment.append(1, 1000L, 1, MemoryRecords.withRecords(1, Compression.NONE, new SimpleRecord("one".getBytes())));
+        assertEquals(1000L, segment.getFirstBatchTimestamp());
+
+        segment.close();
+    }
+
+    private ProducerStateManager newProducerStateManager() throws IOException {
+        return new ProducerStateManager(
+            topicPartition,
+            logDir,
+            5 * 60 * 1000,
```

Comment:
nit: I think `((int) (Duration.ofMinutes(5).toMillis()))` is more readable, WYDT ?